File: accreta-java/java/src/main/java/com/accreta/AggregateSet.java
package com.accreta;

import java.util.OptionalDouble;

/**
 * Mirrors {@code accreta::aggregate_set::AggregateSet}.
 *
 * <p>Rust's {@code set.get::<Sum<f64>>()} is a compile-time-typed lookup; JNI has no generics, so
 * this exposes one typed getter per built-in aggregate instead. Each getter's fallback value
 * matches what the corresponding Rust aggregate would report if it isn't registered on this
 * measure at all — check {@link Schema} up front if you need to tell "not registered" apart from
 * a genuine zero/empty result.
 */
public final class AggregateSet implements AutoCloseable {
    static {
        NativeLibrary.load();
    }

    private long handle;

    /** Package-private: only {@link Engine#queryRange} constructs one. */
    AggregateSet(long handle) {
        this.handle = handle;
    }

    /** {@code Sum<f64>.value()}. */
    public double getSum() {
        checkOpen();
        return nativeGetSum(handle);
    }

    /** {@code Count.value()}. */
    public long getCount() {
        checkOpen();
        return nativeGetCount(handle);
    }

    /**
     * The mean, derived as {@link #getSum()} / {@link #getCount()} ({@code accreta} has no
     * {@code Average} aggregate since 0.2.0). {@code NaN} if the count is zero. Requires both
     * {@code Sum} and {@code Count} to be registered on this measure; otherwise they read as
     * 0.0 / 0 and this returns {@code NaN}.
     */
    public double getAverage() {
        long count = getCount();
        if (count == 0) {
            return Double.NaN;
        }
        return getSum() / (double) count;
    }

    /** {@code Min<f64>.value()} — empty until the first sample, same as the Rust {@code Option}. */
    public OptionalDouble getMin() {
        checkOpen();
        double[] out = new double[1];
        return nativeGetMin(handle, out) ? OptionalDouble.of(out[0]) : OptionalDouble.empty();
    }

    /** {@code Max<f64>.value()}. */
    public OptionalDouble getMax() {
        checkOpen();
        double[] out = new double[1];
        return nativeGetMax(handle, out) ? OptionalDouble.of(out[0]) : OptionalDouble.empty();
    }

    /** {@code TDigest.quantile(q)}. Returns {@code NaN} if TDigest wasn't registered on this measure. */
    public double getQuantile(double q) {
        checkOpen();
        return nativeGetQuantile(handle, q);
    }

    @Override
    public void close() {
        if (handle != 0) {
            nativeDrop(handle);
            handle = 0;
        }
    }

    private void checkOpen() {
        if (handle == 0) {
            throw new IllegalStateException("AggregateSet already closed");
        }
    }

    private static native double nativeGetSum(long handle);

    private static native long nativeGetCount(long handle);

    private static native boolean nativeGetMin(long handle, double[] out);

    private static native boolean nativeGetMax(long handle, double[] out);

    private static native double nativeGetQuantile(long handle, double q);

    private static native void nativeDrop(long handle);
}

File: accreta-java/src/aggregate_set.rs
//! JNI natives backing `com.accreta.AggregateSet`.
//!
//! `AggregateSet::get::<T>()` is a Rust-generic lookup keyed by type; JNI has no equivalent, so
//! it's exposed as one typed getter per built-in aggregate instead (`getSum`, `getCount`, ...).
//! Each returns a sentinel plus a `hasX`/`out-param "found"` story where the aggregate wasn't
//! registered on this measure — see each method's doc below for its exact contract.

use jni::objects::JClass;
use jni::sys::{jboolean, jdouble, jlong, JNI_FALSE, JNI_TRUE};
use jni::JNIEnv;

use accreta::aggregate_set::AggregateSet;
use accreta::aggregates::{Average, Count, Max, Min, Sum, TDigest};

use crate::handles::{borrow, drop_handle};

#[unsafe(no_mangle)]
pub extern "system" fn Java_com_accreta_AggregateSet_nativeDrop(
    _env: JNIEnv,
    _class: JClass,
    handle: jlong,
) {
    unsafe { drop_handle::<AggregateSet>(handle) };
}

/// Sum<f64>. Returns `0.0` (matching the aggregate's own identity value) if `Sum<f64>` wasn't
/// registered on this measure — callers that need to distinguish "zero" from "not registered"
/// should check the schema up front rather than relying on this return value.
#[unsafe(no_mangle)]
pub extern "system" fn Java_com_accreta_AggregateSet_nativeGetSum(
    _env: JNIEnv,
    _class: JClass,
    handle: jlong,
) -> jdouble {
    let set: &AggregateSet = unsafe { borrow(handle) };
    set.get::<Sum<f64>>().map(|s| s.value()).unwrap_or(0.0)
}

/// Count. Returns `0` if `Count` wasn't registered on this measure.
#[unsafe(no_mangle)]
pub extern "system" fn Java_com_accreta_AggregateSet_nativeGetCount(
    _env: JNIEnv,
    _class: JClass,
    handle: jlong,
) -> jlong {
    let set: &AggregateSet = unsafe { borrow(handle) };
    set.get::<Count>().map(|c| c.value()).unwrap_or(0) as jlong
}

/// Average<f64>.sum() — paired with `nativeGetAverageCount` because the Rust side exposes the
/// running sum/count rather than a single `mean()` method (see the `tdigest_quantiles.rs`
/// example: `avg.sum() / avg.count() as f64`). The Java `AggregateSet.getAverage()` wrapper does
/// that division for you.
#[unsafe(no_mangle)]
pub extern "system" fn Java_com_accreta_AggregateSet_nativeGetAverageSum(
    _env: JNIEnv,
    _class: JClass,
    handle: jlong,
) -> jdouble {
    let set: &AggregateSet = unsafe { borrow(handle) };
    set.get::<Average<f64>>().map(|a| a.sum()).unwrap_or(0.0)
}

#[unsafe(no_mangle)]
pub extern "system" fn Java_com_accreta_AggregateSet_nativeGetAverageCount(
    _env: JNIEnv,
    _class: JClass,
    handle: jlong,
) -> jlong {
    let set: &AggregateSet = unsafe { borrow(handle) };
    set.get::<Average<f64>>().map(|a| a.count()).unwrap_or(0) as jlong
}

/// Min<f64>.value(). `Min`'s own value is `Option<f64>` (empty until the first sample), so this
/// writes "did we have a value" into the returned boolean and the value itself into `out[0]` —
/// the Java wrapper turns that into an `OptionalDouble`.
#[unsafe(no_mangle)]
pub extern "system" fn Java_com_accreta_AggregateSet_nativeGetMin(
    env: JNIEnv,
    _class: JClass,
    handle: jlong,
    out: jni::objects::JDoubleArray,
) -> jboolean {
    write_optional(env, handle, out, |set: &AggregateSet| {
        set.get::<Min<f64>>().and_then(|m| m.value())
    })
}

/// Max<f64>.value() — same contract as [`Java_com_accreta_AggregateSet_nativeGetMin`].
#[unsafe(no_mangle)]
pub extern "system" fn Java_com_accreta_AggregateSet_nativeGetMax(
    env: JNIEnv,
    _class: JClass,
    handle: jlong,
    out: jni::objects::JDoubleArray,
) -> jboolean {
    write_optional(env, handle, out, |set: &AggregateSet| {
        set.get::<Max<f64>>().and_then(|m| m.value())
    })
}

fn write_optional(
    env: JNIEnv,
    handle: jlong,
    out: jni::objects::JDoubleArray,
    lookup: impl FnOnce(&AggregateSet) -> Option<f64>,
) -> jboolean {
    let set: &AggregateSet = unsafe { borrow(handle) };
    match lookup(set) {
        Some(value) => {
            env.set_double_array_region(&out, 0, &[value])
                .expect("writing out-param");
            JNI_TRUE
        }
        None => JNI_FALSE,
    }
}

/// TDigest.quantile(q). Returns `f64::NAN` (TDigest's own answer for "no samples yet") if
/// `TDigest` wasn't registered on this measure.
#[unsafe(no_mangle)]
pub extern "system" fn Java_com_accreta_AggregateSet_nativeGetQuantile(
    _env: JNIEnv,
    _class: JClass,
    handle: jlong,
    q: jdouble,
) -> jdouble {
    let set: &AggregateSet = unsafe { borrow(handle) };
    set.get::<TDigest>()
        .map(|d| d.quantile(q))
        .unwrap_or(f64::NAN)
}
